dbms/src/Storages/StorageDistributedShardsOptimizer.cpp

```diff
@@ -0,0 +1,366 @@
+#include <Columns/ColumnConst.h>
```

Comment:
We can make it far more simple and powerful. See the usage of `filterBlockWithQuery`.

Reply:
Can you elaborate? I went through the code, but I don't understand what would be the block parameter. The tricky part is that our sharding key could use a hash function, e.g.,

sharding_key_expr: hash(id) % 2
where_expression: WHERE id = 1

What I see from the code, it wouldn't work for the same reason as ClickHouse doesn't do similar optimization for filtering in MergeTree, in case primary key is `hash(id)`, and we filter `WHERE id = 1`.

Reply:
You're right, let me review it in more detail.